### ISSUE TITLE
parser: redeclare catch parameter with function declaration in non-strict mode should throw a SyntaxError

### ISSUE THREAD

**Issue by Dunqing** [oxc-project/oxc#10056]
[Test262 test](https://github.com/tc39/test262/blob/bc5c14176e2b11a78859571eb693f028c8822458/test/language/statements/try/early-catch-function.js#L26) | [Babel Playground](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=PTAEGEHsAcE8CcCWBzAFgF1ACnASlAEwAMAjABygCCAdgCbwCXoAQgIbzICuANtwHRVeoJGnQBnYQFMxk-ADdJtPgCgQoACqpEEgMaRak0NtDJIC-NUWgARrFDpUh5gGUAIqG6Idk6jNAAzSE46I2p7R1AAGQBJcABRADlnOIDEbkkVVQAqAFo85WlEWgAuUBkdHPR4WByxdFZ0SQBbH3Ra-vQvWubWak6dMRzJdm4a2XhIeDFlAzEdJGhOyGpSgD5lUFAAJUVJHW52BsRl0Eh_CAadVAAFQ5bG-FAAd0QHUFpEeD30UdBLOqsADFgjoltRXHsDvAjidEGF_CCwaA9H1JAAPdAqOGBUoAHw2oBIAGY-CQAKyk0DODpeKk9PpeMSlOIjOxxeATKbKAnRTDGVhU2B9Vho0DszlGc69OySdItPqnc4OJxBOgJVgtCRnC7oK63aH3WSgVjcMSQAmQHQ6ThTULhQyRdFeE2jCH7diKdWaxUsbiWgDWKksyCOCmKBOgqFYMlK0HYMgJ6Fg0EkpWcQvqaPFk2UeRyWWA3IAJK4APIJUvqOIANUokQAqpQq1hcABubkI6ig47wlugADeieqA4JAF9kZdUNhJPhB5tNp3uydJH3-6BR2PlBugA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=&prettier=true&targets=&version=7.27.0&externalPlugins=%40babel%2Fplugin-transform-object-rest-spread%407.25.9&assumptions=%7B%7D) | [OXC Playground](https://playground.oxc.rs/#eNptk9tq20AQhl9lEL1wQuTULYGiQsFxHAiEpMRur3yzXo3sbVezYnaVWLh+oD5HX6yzK8vJRUCgPczhm39n9pnOiuzyEmau6dhstgFGszP49HHyBaZU8r+/cK1401o7hqm1kEw8MHrkZyzHKxLf5dZ40K5EkP/GPSMTlrDuIGwRrhc3YI1G8giVa6kEQ+ni/m42f1jMoTIWJY5EOs/zfEXoTVmAR50H7nIfVMAaKaSV0bnHWpEsfI6KbZcjs2O/ohK9ZtME46iAbysCeMIStVUsbo7AVTBTQW+/y0GNARleTNhCaRh1sB1I3iDUty3paH/zxlWAq+OxlEkBd0GADVWugD8x0+TzeHI1nsAiIcJiQCxgHhlhfmSMtnchqqRg0VFQu/4KTAWKOkCbSo2sSboo14PQ+nfolfUuxnNat+xPouLOaGVt1/Pjq/u1dfq3YBNuBPIZi+jcbJXHAhrFHuM+dI1se7REtiJ5k/PLiP7h5vHhcTn/Ob3/MV3OR2df4+FJl2p0BvsUgrt+cQAdiWGExxt4VRGjNRyS2Yrkyy4ylxX7jFuKP58IsiJwixeZNRSGtdeuwdOmq9fODrvAinzluM6KStTBw0WWCuMYUTRxL08YWqbHNkiL4fDSg3tzbGqRWJr1lMK1rPFWepRESZmVIH0y/uV3QtzfLUUzOe+7L5OkETcllfWJqacIMksotWToZcSuJIbxzkqHl2/6TXIP/HGoNpgkkfxri4uUslZNzycmtSFTmSFfbE929laKjU4yimvnBa8PeDj8BxueYw4=)

```js
// Copyright (C) 2018 André Bargull. All rights reserved.
// This code is governed by the BSD license found in the LICENSE file.

/*---
esid: sec-try-statement-static-semantics-early-errors
description: >
  Redeclaration of CatchParameter with directly nested FunctionDeclaration in function context.
info: |
  13.15.1 Static Semantics: Early Errors

  It is a Syntax Error if any element of the BoundNames of CatchParameter also
  occurs in the LexicallyDeclaredNames of Block.
negative:
  phase: parse
  type: SyntaxError
---*/

$DONOTEVALUATE();

function f() {
  try {
  } catch (e) {
    function e() { }
  }
}
```

### Expect

Both should throw a syntax error in non-strict mode and strict mode

### Actual

Non-strict mode has no error, and strict mode has a syntax error
